File: 235.Lowest-Common-Ancestor-of-a-Binary-Search-Tree.java
// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
//
// algorithms
// Easy (44.32%)
// Total Accepted:    274,537
// Total Submissions: 619,477
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        while (root != null) {
            if (root.val > p.val && root.val > q.val) {
                root = root.left;
            } else if (root.val < p.val && root.val < q.val) {
                root = root.right;
            } else {
                return root;
            }
            
        }
		return null; 
    }
}

File: 235.Lowest-Common-Ancestor-of-a-Binary-Search-Tree.go
// https://leetcode.com/problems/palindrome-linked-list/
//
// algorithms
// Easy (44.65%)
// Total Accepted:    280,819
// Total Submissions: 628,935

package leetcode

/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	for {
		if root.Val == q.Val || root.Val == p.Val {
			return root
		} else if root.Val > p.Val && root.Val > q.Val {
			root = root.Left
		} else if root.Val < p.Val && root.Val < q.Val {
			root = root.Right
		} else {
			return root
		}
	}

	return nil
}
